### PR TITLE
Vqgan training

### PR DESCRIPTION
**Pull request by isamu-isozaki**
This is a draft pr for adding the vqgan training. It's still quite rough around the edges but might be able to do ok after some bug fixes.

### CODE REVIEW

**Comment by isamu-isozaki** [comment]
Tested out on random noise and it runs. I'll try adapting to webdataset on some clusters and see how it does!

**Comment by isamu-isozaki** [comment]
I found https://arxiv.org/abs/2212.03185 thanks to Laion(Ryu) which improves on movq.
The main ideas are
1. Add in perceptual loss from lower layers(which we are already doing)
2. entropy maximization so the codebook usage is 100%

**Comment by isamu-isozaki** [comment]
I'm starting to add the projected gan technique from [here](https://arxiv.org/abs/2111.01007v1). This seems to still have state-of-the-art in quite a few datasets although it is from 2021. The main idea is instead of plugging in images to the generator/discriminator, plugging in timm computed hierarchical features which makes training converge faster.

**Comment by isamu-isozaki** [comment]
Other news is I was finally able to add the imagenet training dataset so I will be testing the movq/spectral norm added f16 pre-trained model soon
